system/src/system_network_manager.cpp

```diff
@@ -609,11 +616,25 @@ void NetworkManager::handleIfLink(if_t iface, const struct if_event* ev) {
         } else if (state_ == State::IP_CONFIGURED || state_ == State::IFACE_LINK_UP) {
             refreshIpState();
         }
+
+        // If the preferred network becomes available, close the cloud connection to force migration to this network
+        if (ConnectionManager::instance()->getPreferredNetwork() == netIfIndex) {
```

Comment:
This should probably be handled outside of network manager with some hook getting called from here.